### PR TITLE
Add benchmarks to CMake

### PR DESCRIPTION
**Pull request by cedricchevalier19**
Small PR that enables building `atomic`, `bytes_and_flops`, `gather` and `policy_performance` benchmarks when `Kokkos_ENABLE_BENCHMARKS` is set.

Small changes are:
- local includes use `"` instead of `<>` to avoid setting `-I.` ,
- renaming some local variables whose names where colliding with others or keyword,
- deleting unneeded function arguments to avoid warnings.

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]

**Review by nmm0 — APPROVED** [review]
Looks good to me!

**Comment by dalg24** [comment]
Seeing
```
fatal error: error in backend: Cannot select: 0x5555645ae330: i64,ch = stacksave 0x5555645ae1e0, core/src/OpenMPTarget/Kokkos_OpenMPTarget_ParallelFor_Team.hpp:150:11 @[ core/src/OpenMPTarget/Kokkos_OpenMPTarget_ParallelFor_Team.hpp:148:1 @[ core/src/OpenMPTarget/Kokkos_OpenMPTarget_ParallelFor_Team.hpp:144:5 ] ]
In function: __omp_offloading_30_4d9e13fa__ZNK6Kokkos4Impl11ParallelForIZN3RunIdLi1ELi1EE3runEiiiiiiiiEUlRKNS0_26OpenMPTargetExecTeamMemberEE_NS_10TeamPolicyIJEEENS_12Experimental12OpenMPTargetEE12execute_implIvEEvv_l144
clang++: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 17.0.3
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/llvm/bin
[ 96%] Building CXX object algorithms/unit_tests/CMakeFiles/Kokkos_AlgorithmsUnitTest_StdSet_Team_H.dir/TestStdAlgorithmsTeamRemoveCopy.cpp.o
clang++: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang++: note: diagnostic msg: /tmp/bench_double-691ca2.cpp
clang++: note: diagnostic msg: /tmp/bench_double-946f6a.cpp
clang++: note: diagnostic msg: /tmp/bench_double-691ca2.sh
clang++: note: diagnostic msg: 

********************
make[2]: *** [benchmarks/bytes_and_flops/CMakeFiles/Kokkos_bytes_and_flops.dir/build.make:76: benchmarks/bytes_and_flops/CMakeFiles/Kokkos_bytes_and_flops.dir/bench_double.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:5285: benchmarks/bytes_and_flops/CMakeFiles/Kokkos_bytes_and_flops.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
in the OpenMPTarget build

**Comment by rgayatri23** [comment]
@cedricchevalier19 - Can you provide me with your cmake configuration line, I have been using this command
```bash
cmake -DCMAKE_CXX_COMPILER=clang++ -DKokkos_ENABLE_OPENMPTARGET=ON -DCMAKE_BUILD_TYPE=RelWithDebugInfo -DKokkos_ARCH_AMPERE80=ON -DKokkos_ENABLE_BENCHMARKS=ON ../
```

And this leads to ICE even when disabling all benchmarks. I tested this with clang/16 and clang/17.0.6 on NVIDIA A100. 

Just wanted to see if there is any difference in the options used by you.

**Comment by rgayatri23** [comment]
Ok I see what the issue is, any `debug` symbol leads to an ICE. This is something that has been noticed even earlier. So the `Release` version works fine except for 2 benchmarks. I am not able to push to your branch but here is the diff that should resolve the issue

```bash
diff --git a/benchmarks/CMakeLists.txt b/benchmarks/CMakeLists.txt
index 4faeb0a20..43f8ae808 100644
--- a/benchmarks/CMakeLists.txt
+++ b/benchmarks/CMakeLists.txt
@@ -1,7 +1,11 @@
 KOKKOS_ADD_BENCHMARK_DIRECTORIES(atomic)
-KOKKOS_ADD_BENCHMARK_DIRECTORIES(bytes_and_flops)
 KOKKOS_ADD_BENCHMARK_DIRECTORIES(gather)
 KOKKOS_ADD_BENCHMARK_DIRECTORIES(gups)
 KOKKOS_ADD_BENCHMARK_DIRECTORIES(launch_latency)
-KOKKOS_ADD_BENCHMARK_DIRECTORIES(policy_performance)
 KOKKOS_ADD_BENCHMARK_DIRECTORIES(stream)
+
+#FIXME_OPENMPTARGET - These two benchmarks cause ICE. Commenting them for now but a deeper analysis on the cause and a possible fix will follow.
+IF(NOT Kokkos_ENABLE_OPENMPTARGET)
+KOKKOS_ADD_BENCHMARK_DIRECTORIES(policy_performance)
+KOKKOS_ADD_BENCHMARK_DIRECTORIES(bytes_and_flops)
+ENDIF()
```

File: benchmarks/CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,22 @@
+if (Kokkos_ENABLE_OPENMPTARGET AND CMAKE_BUILD_TYPE!="Release" AND Kokkos_ENABLE_BENCHMARKS)
```
Comment (dalg24) [inline benchmarks/CMakeLists.txt:1]: > `CMAKE_BUILD_TYPE!="Release"`

I don't think that is valid cmake

File: benchmarks/CMakeLists.txt [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
+if (NOT Kokkos_ENABLE_OPENMPTARGET)
     KOKKOS_ADD_BENCHMARK_DIRECTORIES(policy_performance)
     KOKKOS_ADD_BENCHMARK_DIRECTORIES(bytes_and_flops)
-ENDIF()
\ No newline at end of file
+else ()
+    # FIXME_OPENMPTARGET - These two benchmarks cause ICE.
+    # Commenting them for now but a deeper analysis on the cause and a possible fix will follow.
+    if (Kokkos_ENABLE_BENCHMARKS)
+        message(WARNING "OpenMPTarget is enabled: disabling 2 benchmarks to avoid ICE.")
```
Comment (dalg24) [inline benchmarks/CMakeLists.txt:20]: I would't raise a warning for this.  I think the previous version was good enough.

File: benchmarks/CMakeLists.txt [hunk 3]
```diff
@@ -1,11 +1,22 @@
...
+    endif ()
+endif ()
```
Comment (dalg24) [inline benchmarks/CMakeLists.txt:22]: Add an empty new line

File: benchmarks/CMakeLists.txt (hunk 2 above)
Comment (cedricchevalier19) [reply benchmarks/CMakeLists.txt]: Ok, I revert the last commit, and rebase to have an updated SYCL test.

**Review by masterleinad — APPROVED** [review]

File: benchmarks/bytes_and_flops/bench_unroll_stride.hpp [hunk 4]
```diff
@@ -87,25 +87,25 @@ struct Run<Scalar, UNROLL, STRIDE> {
                     C(n, i, 0) = a1;
 #endif
 #if (UNROLL == 2)
-                    C(n, i, 0) = a1 + a2;
+                    C(n, iter, 0) = a1 + a2;
```
Comment (dalg24) [inline benchmarks/bytes_and_flops/bench_unroll_stride.hpp:90]: What about L87?
Also that does not look right to me.  It gives out-of-bounds accesses if the number of iterations exceeds the view extent in this dimension.  Also it changes the meaning of the code.

Comment (crtrott) [reply benchmarks/bytes_and_flops/bench_unroll_stride.hpp:90]: yeah this all needs to stay `i` that said the rename of the shadowed variable is probably a good idea :-)

File: benchmarks/bytes_and_flops/main.cpp [hunk 5]
```diff
@@ -86,7 +86,7 @@ int main(int argc, char* argv[]) {
     printf("D must be one of 1,2,4,8,16,32\n");
     return 0;
   }
-  if ((P < 1) && (P > 2)) {
+  if ((P < 1) || (P > 4)) {
```
Comment (crtrott) [inline benchmarks/bytes_and_flops/main.cpp:89]: haha that condition above is funny :-)

**Review by crtrott — DISMISSED** [review]
Yeah the change in the loop as Damien found is wrong - needs to be `i` other than that looks good.

File: benchmarks/bytes_and_flops/bench_unroll_stride.hpp (hunk 4 above)
Comment (cedricchevalier19) [reply benchmarks/bytes_and_flops/bench_unroll_stride.hpp:90]: Oups, very sorry but thank you for this good catch @dalg24.

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Willing to ignore Clang+CUDA and OpenMPTarget builds that were interrupted and one of the HIP build that had (unrelated) runtime issues.

File: benchmarks/CMakeLists.txt [hunk 6]
```diff
@@ -1,3 +1,12 @@
+#FIXME_OPENMPTARGET - compiling in debug mode causes ICE.
```
Comment (dalg24) [inline benchmarks/CMakeLists.txt:1]: @rgayatri23 do you want this to be guarded?

Comment (rgayatri23) [reply benchmarks/CMakeLists.txt:1]: Yeah that would be a good idea to protect it.

**Review by crtrott — APPROVED** [review]
I think this is ok without an extra guard. Will remind us that there is actually an issue.

**Pull request merged**